library/alloc/src/sync.rs

```diff
@@ -500,6 +500,62 @@ impl<T> Arc<T> {
             Ok(elem)
         }
     }
+
+    /// Returns the inner value, if the `Arc` has exactly one strong reference.
+    ///
+    /// Otherwise, [`None`] is returned and the `Arc` is dropped.
+    ///
+    /// This will succeed even if there are outstanding weak references.
+    ///
+    /// If `unwrap_or_drop` is called on every clone of this `Arc`,
+    /// it is guaranteed that exactly one of the calls returns the inner value.
+    /// The similar expression `Arc::try_unwrap(this).ok()` does not
+    /// offer this guarantee.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #![feature(unwrap_or_drop)]
+    ///
+    /// use std::sync::Arc;
+    ///
+    /// let x = Arc::new(3);
+    /// let y = Arc::clone(&x);
+    ///
+    /// let x_unwrap_thread = std::thread::spawn(|| Arc::unwrap_or_drop(x));
+    /// let y_unwrap_thread = std::thread::spawn(|| Arc::unwrap_or_drop(y));
+    ///
+    /// let x_unwrapped_value = x_unwrap_thread.join().unwrap();
+    /// let y_unwrapped_value = y_unwrap_thread.join().unwrap();
+    ///
+    /// assert!(matches!(
+    ///     (x_unwrapped_value, y_unwrapped_value),
+    ///     (None, Some(3)) | (Some(3), None)
+    /// ));
+    /// ```
+    #[inline]
+    #[unstable(feature = "unwrap_or_drop", issue = "none")] // FIXME: add issue
+    // FIXME: should this copy all/some of the comments from drop and drop_slow?
+    pub fn unwrap_or_drop(this: Self) -> Option<T> {
+        // following the implementation of `drop` (and `drop_slow`)
+        let mut this = core::mem::ManuallyDrop::new(this);
+
+        if this.inner().strong.fetch_sub(1, Release) != 1 {
+            return None;
+        }
+
+        acquire!(this.inner().strong);
+
+        // FIXME: should the part below this be moved into a seperate #[inline(never)]
+        // function, like it's done with drop_slow in drop?
+
+        // using `ptr::read` where `drop_slow` was using `ptr::drop_in_place`
+        let inner = unsafe { ptr::read(Self::get_mut_unchecked(&mut this)) };
```

Comment:
I forgot: Please write a `// SAFETY:` comment for why this is safe here. It doesn't have to be very long, just explains why the invariants are checked. :) [Here is an example from `core`](https://github.com/rust-lang/rust/blob/48717b6f3ce661d2a0d64f7bdfdfb5fd3484ee5b/library/core/src/slice/mod.rs#L72).

Reply:
The situation is a bit more complex. It is safe because it is doing the same things that the `drop` implementation does, with the difference of doing `ptr::read` instead of `ptr::drop_in_place`, but those two operations are applicable in pretty much exactly the same situations. The reason why every step of `drop` is safe includes some really really really long comments, hence my general question, as stated in the `FIXME` in line 538: “should I copy [...] the comments from drop?”

Reply:
What I've done before is say something along the lines of "this is safe for the same reason the `drop` impl is safe; see that for more info".

Reply:
Sorry I didn't answer earlier, I didn't get the notification for this comment.

The comments in the `drop` implementation should not change very much, except for improving them so referencing them is probably okay.

You can add a line about the differences between `ptr::read` and `ptr::drop_in_place` and why the former must be used instead of the later below and it should be good.